### ISSUE TITLE
wxIMAGE_QUALITY_HIGH not working with MSVS 2022

### ISSUE THREAD

**Issue by dkulp** [wxWidgets/wxWidgets#22761]
**Describe the bug**

The xLights team has moved up to the latest wxWidgets master code (plus some fixes).  However, we are having some problems with scaled images on Windows, but only using optimized builds with Visual Studio.  Debug builds are fine as are the Mac and Linux builds.    The code we are using is similar to:

```
wxBitmapButton* LayoutPanel::CreateModelButton(const char *data[]) {
    wxImage image(data);
    wxImage img24 = image.Scale(24, 24, wxIMAGE_QUALITY_HIGH);
    wxBitmapBundle bmp = wxBitmapBundle::FromBitmaps(img24, image);
    return new wxlBitmapButton(panel, bmp);
 }
```
Basically, the images (which are 48x48) are scaled down to 24x24 and a wxBitmapBundle created with the 24x24 and 48x48 images and a wxBitmapButton created using that.    On Windows optimized builds on non-HiDPI screens, the resulting image is black.   (not sure on HiDPI screens)    The workaround we are using is to use wxIMAGE_QUALITY_BICUBIC on Windows.  I've attached sample images in a zip. (github doesn't support attaching xpm's directly).   data is the xpm data.

See:
https://github.com/smeighan/xLights/issues/3388
https://github.com/smeighan/xLights/issues/3397

**Expected vs observed behaviour**
Scaled images display correctly.

**Patch or snippet allowing to reproduce the problem**

**Platform and version information**
 - wxWidgets 3.3 (master branch as of Aug 24)
 - wxWidgets port you use: wxMSW
 - OS and its version: Windows 10
  
[images.zip](https://github.com/wxWidgets/wxWidgets/files/9452603/images.zip)

**Comment by PBfordev** [comment]
I can confirm this, it seems unrelated to bitmap bundle or XPM and can be reproduced in the image sample with this trivial patch
```diff
diff --git a/samples/image/canvas.cpp b/samples/image/canvas.cpp
index 9a634fed4f..cb62d367e4 100644
--- a/samples/image/canvas.cpp
+++ b/samples/image/canvas.cpp
@@ -96,7 +96,7 @@ MyCanvas::MyCanvas( wxWindow *parent, wxWindowID id,
     }
     else
     {
-        my_horse_png = wxBitmap( image );
+        my_horse_png = image.Scale(128, 128, wxIMAGE_QUALITY_HIGH);
     }
 
     if ( !image.LoadFile( dir + "toucan.png") )
```
Run the patched sample in the Release mode and see that the picture for "PNG Handler" is just a black square. The picture is displayed correctly in the Debug mode.

Tested the current master with MSVS 2022 Version 17.3.0, the issue manifests with wxWidgets built both by NMake makefile and CMake-generated MSVS solution.

Could not reproduce with GCC (12.2 from MSYS2 package mingw-w64-x86_64-toolchain) using CMake-generated makefile.

**Comment by dkulp** [comment]
The only change in that method in the last 5 years was changing 

```
if (sum_a)
```
to 
```
if (sum_a != 0)
```

Any chance you can check if reversing that would fix it?   Maybe "if (sum_a > 0.0f)" or similar?

**Comment by trogmaniac** [comment]
> **EDIT 2** It seems that declaring volatile any single variable of those five makes the bug disappear but fixes like this look like alchemy, not programming

Reminds me of the wxQuantize bug from a few years ago:
https://github.com/wxWidgets/wxWidgets/issues/17764

**Comment by PBfordev** [comment]
> The only change in that method in the last 5 years was changing ...  Any chance you can check if reversing that would fix it? Maybe "if (sum_a > 0.0f)" or similar?

I tried the old variant and your suggestion (also without "f"), but unfortunately with no effect.

**Comment by vadz** [comment]
I can't reproduce this with MSVS 2019 (16.8.2) using the following diff:
```diff
diff --git a/tests/image/image.cpp b/tests/image/image.cpp
index 00ac928a03..1fc046c6f7 100644
--- a/tests/image/image.cpp
+++ b/tests/image/image.cpp
@@ -2333,6 +2333,19 @@ static void LoadMalformedImageWithException(const wxString& file,
 #endif // SIZEOF_VOID_P == 8
 }
 
+// Trivial test for high quality resampling.
+TEST_CASE("wxImage::Scale/HighQuality", "[image]")
+{
+    wxImage img;
+    REQUIRE( img.LoadFile("horse.bmp") );
+
+    CHECK( img.GetSize() == wxSize(200, 200) );
+
+    REQUIRE_NOTHROW( img.Rescale(300, 300, wxIMAGE_QUALITY_HIGH) );
+
+    CHECK( img.GetSize() == wxSize(300, 300) );
+}
+
 // This can be used to test loading an arbitrary image file by setting the
 // environment variable WX_TEST_IMAGE_PATH to point to it.
 TEST_CASE("wxImage::LoadPath", "[.]")
```
so it's difficult for me to test any potential fixes...

BTW, do you see it in 32 or 64-bit builds or both?

**Comment by dkulp** [comment]
@vadz    The resize works so testing the size won't  reveal anything.   This issue is that the resulting resized image is black.

**Comment by PBfordev** [comment]
I cannot reproduce the issue (using modified image sample I posted) with 64-bit build in MSVS 2019 (v16.11.18) nor 32-bit build in MSVS 2022 (v17.3.0) 32-bit.

The issue is still present in 64-bit build in MSVS 2022 even after upgrade from v17.3.0 to 17.3.3.

**Comment by vadz** [comment]
I really should slow down when reading the issues, sorry. I thought it crashed (as in the issue with wxQuantize()) for some reason...

But I can confirm what @PBfordev has already written: it works fine with MSVS 2019, so it looks like a 2022 regression.

FWIW I would be fine with adding `volatile` or doing any other nonsensical change to make this work. Looking at the code, perhaps we should rewrite it to stop using `src_pixel_index` and just use pointers instead, this might result in better code and also avoid the buggy optimization.

**Comment by PBfordev** [comment]
As 3.3.1 release date is slowly approaching...

This stupid hack seems to make the bug go away:
```diff
diff --git a/src/common/image.cpp b/src/common/image.cpp
index 76f2910987..badd6b2526 100644
--- a/src/common/image.cpp
+++ b/src/common/image.cpp
@@ -660,7 +660,10 @@ wxImage wxImage::ResampleBox(int width, int height) const
     }
 
     int averaged_pixels, src_pixel_index;
-    double sum_r, sum_g, sum_b, sum_a;
+    double sum_r, sum_g, sum_b;
+    // sum_a declared volatile to avoid MSVS 2022 optimization bug,
+    // see https://github.com/wxWidgets/wxWidgets/issues/22761
+    volatile double sum_a;
 
     for ( int y = 0; y < height; y++ )         // Destination image - Y direction
     {
```
FWIW, `bench_gui ShrinkHighQuality` shows the patched MSVS 2022 build has similar performance as unpatched MSVS 2019 (~125 us avg) while the bugged MSVS 2022 is almost three times faster (~50 us).

If nothing else appears before the release deadline, perhaps it is better to use something like this then keep the bug unfixed. Of course, there is no guarantee that the next MSVC release will not "improve" the optimizer again and reintroduce the bug.

**Comment by MaartenBent** [comment]
I found another fix that seems to work for me, calculate `averaged_pixels` instead of incrementing it:
```diff
 src/common/image.cpp | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/src/common/image.cpp b/src/common/image.cpp
index 76f2910987..8cc3cd6463 100644
--- a/src/common/image.cpp
+++ b/src/common/image.cpp
@@ -673,7 +673,7 @@ wxImage wxImage::ResampleBox(int width, int height) const
             const BoxPrecalc& hPrecalc = hPrecalcs[x];
 
             // Box of pixels to average
-            averaged_pixels = 0;
+            averaged_pixels = (vPrecalc.boxEnd - vPrecalc.boxStart + 1) * (hPrecalc.boxEnd - hPrecalc.boxStart + 1);
             sum_r = sum_g = sum_b = sum_a = 0.0;
 
             for ( int j = vPrecalc.boxStart; j <= vPrecalc.boxEnd; ++j )
@@ -696,8 +696,6 @@ wxImage wxImage::ResampleBox(int width, int height) const
                         sum_g += src_data[src_pixel_index * 3 + 1];
                         sum_b += src_data[src_pixel_index * 3 + 2];
                     }
-
-                    averaged_pixels++;
                 }
             }
```

**Comment by vadz** [comment]
The last fix seems the best to me as it also could be slightly faster, in theory. I'll apply it if there are no objections or better suggestions, thanks!

**Comment by vadz** [comment]
Should be fixed by the commits above.

**Comment by petebannister** [comment]
I've encountered this also with wx 3.1.4.  It works with VS 17.5.3 but not 17.5.4 and only on release/x64.  release/x86 is fine.   Using WPO - its possible without LTCG it would work but did not try that.

**Comment by vadz** [comment]
The important question is whether you still see the problem in the latest 3.2/master branches.

**Comment by petebannister** [comment]
It works in 3.2.2.1!  I was trying to avoid upgrading right now due to resolving any knock-on effects (there were many, in particular the owner drawn menu issue which was not resolved in 3.2.2.1 so I have to re-implement a load of that to fix for my application).  Interestingly bing chat suggested that 3.1.6 would fix it but it didn't.  But it did direct me to this thread which was very useful.

I thought it was important to mention that the compiler went wrong between versions 17.5.3 and 17.5.4.  17.6.2 is also broken in this regard (Microsoft / GPT-4 if you're listening in..)

**Comment by vadz** [comment]
Sorry, what is the "owner drawn menu issue"?

**Comment by petebannister** [comment]
owner drawn menu issue was #23230 .  As a workaround I have derived from wxMenuItem and simply don't set the bitmaps but then override the measureitem / drawitem to render it.

**Comment by vadz** [comment]
The fix for #23230 is not in 3.2.2, but will be in 3.2.3. Please try applying 1e98812a56a494dce60384ba01b7c0a77b4fd651 to your version and check if it works.

**Comment by petebannister** [comment]
Thank you - my workaround is working fine for now - I'll try reverting that once 3.2.3 is out.  Its a bit painful to apply patches since wx is built by conan package manager in my setup
